### ISSUE TITLE
prjconf `Release:` breaks debs’ debian revision

### ISSUE THREAD

**Issue by drmuey** [openSUSE/obs-build#695]
**TL;DR**:

Doing the classic `Release: %%{?release_prefix}.<CI_CNT>.<B_CNT>` in prjconf breaks debian revision.

1. set `OBS-DCH-RELEASE: 1`:
   * :green_circle: deb is VERSION+release_prefix.<CI_CNT>.<B_CNT>
   * :red_circle: rpm is VERSION-release_prefix
2. add `Release: %%{?release_prefix}.<CI_CNT>.<B_CNT>` in prjconf
   * :red_circle: deb is VERSION+release_prefix
   * :green_circle: rpm is VERSION-release_prefix.<CI_CNT>.<B_CNT>

**Details**:

This problem ➜ https://forums.opensuse.org/showthread.php/554631

I believe is caused by this line being too greedy and clearing Release completely:

https://github.com/openSUSE/obs-build/blob/d8665ae4d182527dec454328222491aa86de522b/build-recipe-dsc#L52

**Comment by ColMelvin** [comment]
To handle macros with macros embedded in them, e.g. `%{expand: %%{mymacro}%%{mysuffix}}`, there will need to be a more complicated regex.  An example might be:

```
s/
    (?(DEFINE)
        (?<MACRO> %\{ (?&EXPANSION) \} | %(?&VARIABLE) | %\*\* | %[*#] )
        (?<EXPANSION> [^%]+ | [^%]* (?: (?&MACRO) [^%]* )+ )
        (?<VARIABLE> \w+ )
    )
    (?&MACRO)\.?
//xg;
```

I don't think this solution will work well with parameterized macro calls (e.g `%mymacro 5`), so further iteration may be required.

**Comment by drmuey** [comment]
@ColMelvin and I were able to verify that this is indeed the problem 👍

We get it to work w/ a simplified regex that will only work on the classic prjconf `Release: %%{?release_prefix}.<CI_CNT>.<B_CNT>` (lots of escape issues and special character oddness leyt along complexities of macro format).


First, thi sgot us the correct `DEB_RELEASE`:

```
--- build-recipe-dsc.orig	2021-06-01 12:49:32.606955951 -0500
+++ build-recipe-dsc	2021-06-01 14:42:17.490308842 -0500
@@ -49,7 +49,10 @@
     done
     # remove rpm macros (everything after "%")
     # they are not evaluated by the Debian build process
-    DEB_RELEASE=`sed 's/%.*$//' <<< $RELEASE`
+    DEB_RELEASE=`sed 's/%..release_prefix..//' <<< $RELEASE`
     OBS_DCH_RELEASE=""
 
     if test -n "$DEB_TRANSFORM" ; then 
     ```
     
Some debuggery showed that `DEB_RELEASE` correctly had `%{?release_prefix}.` removed but we still had an error like `version '1.0-53+%{?release_prefix}.53.2.cpanel' is invalid: version number`.

That seemed to indicate something was using `RELEASE` when it should be using `DEB_RELEASE` (a second bug maybe?) so we tried this patch:

```
--- build-recipe-dsc.orig	2021-06-01 12:49:32.606955951 -0500
+++ build-recipe-dsc	2021-06-01 14:56:04.743476406 -0500
@@ -49,7 +49,11 @@
     done
     # remove rpm macros (everything after "%")
     # they are not evaluated by the Debian build process
-    DEB_RELEASE=`sed 's/%.*$//' <<< $RELEASE`
+    RELEASE=`sed 's/%..release_prefix..//' <<< $RELEASE`
+    DEB_RELEASE=$RELEASE
+
     OBS_DCH_RELEASE=""
     if test -n "$DEB_TRANSFORM" ; then 
```
and it worked!

by “worked” I mean:

* version is 1.0, `release_prefix` is 53 (in debian/ by virtue of the changelog)
* projconf had `Release: %%{?release_prefix}.<CI_CNT>.<B_CNT>`
* `.dsc` had `OBS-DCH-RELEASE: 1`
* when they built:
    * 🟢 RPM was v1.0-53.53.4
    * 🟢 deb was v1.0-53+53.2

HTH, thanks!

**Comment by drmuey** [comment]
Perhaps Iteration 1 is: support the standard `Release: %%{?release_prefix}.<CI_CNT>.<B_CNT>` for debs

Then Iteration 2: make the regexp smarter

**Comment by drmuey** [comment]
PR #698
